### PR TITLE
Adding Simulation and MultistateSimulation reporter

### PR DESCRIPTION
**Pull request by wiederm**
## Description

For a given MCMC sequence we want to obtain statistics, snapshots (set of specified indices) and log properties (e.g. density) for each MCMC sampler. The multistate sampler needs its reporter to provide the potential energies and state indices to calculate free energies. Additionally, the multistate sampler needs to save regular checkpoint files from which it is possible to restart the MCMC chain.

Note: This PR also contains the improved PRNG implementation. Now, the random number stream is a broadcaster in the sampler state of each MCMC move. This was necessary to ensure that random number streams were separated, which became an issue during this PR since trajectories were synchronizing.

## Todos
  - [x] BaseReporter class
  - [X] LangevinReporter
  - [x] MultistateReporter

## Status
- [ ] Ready to go

### CODE REVIEW

File: chiron/states.py [hunk 1]
```diff
@@ -22,6 +23,7 @@ class SamplerState:
     def __init__(
         self,
         x0: unit.Quantity,
+        random_seed: random.PRNGKey,
```
Comment (chrisiacovella) [inline chiron/states.py:26]: random.PRNGKey is the current state of the sampler...hence I don't think `random_seed` is the current name. This should probably be called `current_key` or `current_PRNG_key`.  

I think the sampler state `init`  function should accept two things:
- random_seed that would be used to generate the first key. 
- current_key which would just set self._current_key. 
where if current_key is defined, we will ignore the random_seed if set, or throw an error.

File: chiron/reporters.py [hunk 2]
```diff
@@ -99,11 +151,122 @@ def get_property(self, name: str):
...
+        super().__init__(file_path=self.file_path, buffer_size=buffer_size)
+
+    @classmethod
+    def get_name(cls):
+        return cls._name
+
+
+class LangevinDynamicsReporter(_SimulationReporter):
+    _name = "langevin_reporter"
+    _default_properties = ["trajectory", "box_vectors", "potential_energy", "time"]
```
Comment (chrisiacovella) [inline chiron/reporters.py:236]: Non-blocking, but making note that we probably want to in the future add in the computed temperature.

**Review by chrisiacovella — CHANGES_REQUESTED** [review]
A few things about how we are implementing the RNG in sampler_state and I think we need to only be opening the HDF5 when writing data (and then closing when done) to avoid corrupt data and allow examination during a run

File: chiron/utils.py [hunk 3]
```diff
@@ -1,5 +1,25 @@
...
+    def __init__(self) -> None:
+        pass
+
+    @classmethod
+    def set_seed(cls, seed: int) -> None:
+        cls._seed = seed
+        cls._key = random.PRNGKey(seed)
+
+    @classmethod
+    def get_random_key(cls) -> int:
```
[the remark this answers is not in the document]
Comment (wiederm) [reply chiron/utils.py:19]: I agree that there might be a better way to do this. I wanted to have a solution that (a) requires only a single random seed that the user provides for any number of SamplerState and (b) passes the responsibility for the PRN stream to the SamplerState (which then can live on different machines). This lets us reproduce (if so desired) the random numbers consistently, but also ensures that each SamplerState has an unique PRN stream.

File: chiron/states.py (hunk 1 above)
Comment (wiederm) [reply chiron/states.py:26]: According to https://jax.readthedocs.io/en/latest/jax-101/05-random-numbers.html, `seed` and `key` are synonyms :-)  
But, I agree that it is clearer if we call it `current_PRNG_key`.

I would prefer to only pass `current_PRNG_key` and use it to generate the first key of the new stream. Is there a use case in which we would like to manually set the `current_key`?

File: chiron/integrators.py [hunk 4]
```diff
@@ -41,8 +43,9 @@ def __init__(
             Frequency of saving the simulation data. Default is 100.
```
[the remark this answers is not in the document]
Comment (wiederm) [reply chiron/integrators.py:43]: I think the integrator/move (i.e., whatever class does the calculation that the reporter should log) is responsible for managing the frequency of logging (in other words, the reporter shouldn't know about what it logs and how frequently it does so at initialization). The reporter on the other hand should be responsible of the buffering (`buffer_size`), which controls how often we write to disk.

> Alternatively, should save traj in memory be in the reporter? I.e., it just doesn't clear out the trajectory information when saving to disk? 
That's a great idea! Can we leave it for this PR like it is but open an issue that describes what should be changed?

File: chiron/mcmc.py [hunk 5]
```diff
@@ -16,20 +21,24 @@ def __init__(self, nr_of_moves: int, seed: int):
...
 
 
 class LangevinDynamicsMove(MCMCMove):
     def __init__(
         self,
         stepsize=1.0 * unit.femtoseconds,
         collision_rate=1.0 / unit.picoseconds,
-        simulation_reporter: Optional[SimulationReporter] = None,
+        reporter: Optional[LangevinDynamicsReporter] = None,
         nr_of_steps=1_000,
```
[the remark this answers is not in the document]
Comment (wiederm) [reply chiron/mcmc.py:41]: Great catch!

File: chiron/integrators.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (wiederm) [reply chiron/integrators.py:43]: But I am unsure if the current way (passing it to `__init__`) is the right idea. I suggest to change this and pass it in the `run` method. What do you think?

File: chiron/mcmc.py [hunk 6]
```diff
@@ -420,7 +411,7 @@ def apply(
             log.debug(
                 f"Move accepted. Energy change: {delta_energy:.3f} kT. Number of accepted moves: {self.n_accepted}."
             )
-            reporter.report(
+            self.reporter.report(
```
[the remark this answers is not in the document]
Comment (wiederm) [reply chiron/mcmc.py:414]: Done! Even though I think we want to log every successful MCMove, so the `report_frequency` should be 1

File: chiron/reporters.py [hunk 7]
```diff
@@ -5,28 +5,58 @@
...
+        self.file_path = file_path
+        self.workdir = file_path.parent
+        log.info(f"Writing simulation data to {self.file_path}")
+
         self.buffer_size = buffer_size
-        self.topology = topology
         self.buffer = {}
-        self.h5file = h5py.File(filename, "a")
-        log.info(f"Writing simulation data to {filename}")
+        self.h5file = h5py.File(self.file_path, "a")
```
[the remark this answers is not in the document]
Comment (wiederm) [reply chiron/reporters.py:51]: That is a good point! I have switched to context manager for the read-and-write operations.

File: chiron/integrators.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (chrisiacovella) [reply chiron/integrators.py:43]: Makes sense!

File: chiron/mcmc.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (chrisiacovella) [reply chiron/mcmc.py:414]: I'm sure there are cases where we would not want to log every move (certainly not in a lot of the simple test cases), but setting default to be 1 is probably reasonable

File: chiron/reporters.py [hunk 8]
```diff
@@ -5,28 +5,58 @@
...
+
+
+class _SimulationReporter:
+    def __init__(self, file_path: pathlib.Path, buffer_size: int = 10):
         """
         Initialize the SimulationReporter.
 
         Parameters
         ----------
         filename : str
```
[the remark this answers is not in the document]
Comment (chrisiacovella) [reply chiron/reporters.py:40]: I definitely missed this was a private class and the setup is handled in the public class.

File: chiron/reporters.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (chrisiacovella) [reply chiron/reporters.py:51]: I'm not seeing that change to the context manager in the code, did you forget to commit that change?

**Review by chrisiacovella — APPROVED** [review]
I think this looks good and can be merged.
